File: node/cmd/reader_core_bench.js
#!/usr/bin/env node
// Node.js reader-core benchmark command.

import { readFileSync } from 'node:fs';
import { performance } from 'node:perf_hooks';
import {
  DirectoryReader,
  FileReader,
  SdJournalEnumerateAvailableData,
  SdJournalNext,
  SdJournalOpenDirectory,
  SdJournalOpenFiles,
  SdJournalPrevious,
  SdJournalRestartData,
} from '../src/index.js';

const MASK64 = (1n << 64n) - 1n;

class Counts {
  constructor() {
    this.records = 0;
    this.fields = 0;
    this.bytes = 0;
    this.checksum = 0n;
  }

  addPayload(payload) {
    this.fields += 1;
    this.bytes += payload.length;
    this.checksum = checksumPayload(this.checksum, payload);
  }

  addRecordMarker(value) {
    this.records += 1;
    this.checksum = rotateLeft(this.checksum, 7n) ^ BigInt(value);
    this.checksum &= MASK64;
  }
}

function rotateLeft(value, shift) {
  value &= MASK64;
  return ((value << shift) | (value >> (64n - shift))) & MASK64;
}

function checksumPayload(checksum, payload) {
  checksum = rotateLeft(checksum, 5n) ^ BigInt(payload.length);
  if (payload.length > 0) {
    checksum ^= BigInt(payload[0]) << 8n;
    checksum ^= BigInt(payload[payload.length - 1]);
  }
  return checksum & MASK64;
}

function processStatusKb() {
  let status;
  try {
    status = readFileSync('/proc/self/status', 'utf8');
  } catch {
    return {};
  }
  const wanted = new Set([
    'VmSize', 'VmPeak', 'VmRSS', 'VmHWM', 'RssAnon', 'RssFile',
    'RssShmem', 'VmData', 'VmStk', 'VmExe', 'VmLib', 'VmPTE',
  ]);
  const out = {};
  for (const line of status.split('\n')) {
    const idx = line.indexOf(':');
    if (idx < 0) continue;
    const key = line.slice(0, idx);
    if (!wanted.has(key)) continue;
    const parts = line.slice(idx + 1).trim().split(/\s+/);
    if (parts.length > 0) {
      const value = Number(parts[0]);
      if (Number.isFinite(value)) out[`${key}_kb`] = value;
    }
  }
  return out;
}

function advance(reader, direction) {
  return direction === 'backward' ? reader.previous() : reader.next();
}

function openSdkReader(inputs, surface) {
  if (surface === 'file') {
    if (inputs.length !== 1) throw new Error('file surface requires exactly one --input');
    return FileReader.open(inputs[0]);
  }
  if (surface === 'directory') {
    if (inputs.length !== 1) throw new Error('directory surface requires exactly one --input');
    return DirectoryReader.open(inputs[0]);
  }
  if (surface === 'open-files') {
    return DirectoryReader.openFiles(inputs);
  }
  throw new Error(`invalid surface: ${surface}`);
}

function seekReader(reader, direction) {
  if (direction === 'backward') reader.seekTail();
  else reader.seekHead();
}

function readSdk(inputs, surface, mode, direction) {
  const reader = openSdkReader(inputs, surface);
  try {
    seekReader(reader, direction);
    const counts = new Counts();
    while (advance(reader, direction)) {
      if (mode === 'sdk-entry') {
        const entry = reader.getEntry();
        counts.addRecordMarker(entry.realtime);
        for (const payload of entry.payloads) counts.addPayload(payload);
      } else if (mode === 'sdk-payloads') {
        counts.addRecordMarker(reader.getRealtimeUsec());
        reader.visitEntryPayloads((payload) => counts.addPayload(payload));
      } else {
        throw new Error(`invalid SDK mode: ${mode}`);
      }
    }
    return counts;
  } finally {
    reader.close();
  }
}

function openFacade(inputs, surface) {
  if (surface === 'file' || surface === 'open-files') return SdJournalOpenFiles(inputs, 0);
  if (surface === 'directory') {
    if (inputs.length !== 1) throw new Error('directory surface requires exactly one --input');
    return SdJournalOpenDirectory(inputs[0], 0);
  }
  throw new Error(`invalid facade surface: ${surface}`);
}

function readFacade(inputs, surface, mode, direction) {
  const journal = openFacade(inputs, surface);
  try {
    seekReader(journal, direction);
    const counts = new Counts();
    for (;;) {
      const advanced = direction === 'backward' ? SdJournalPrevious(journal) : SdJournalNext(journal);
      if (advanced === 0) break;
      if (mode === 'facade-next') {
        counts.addRecordMarker(journal.getRealtimeUsec());
      } else if (mode === 'facade-data') {
        counts.addRecordMarker(journal.getRealtimeUsec());
        SdJournalRestartData(journal);
        for (;;) {
          const payload = SdJournalEnumerateAvailableData(journal);
          if (payload === null) break;
          counts.addPayload(payload);
        }
      } else {
        throw new Error(`invalid facade mode: ${mode}`);
      }
    }
    return counts;
  } finally {
    journal.close();
  }
}

function parseArgs(argv) {
  const args = Object.assign(Object.create(null), {
    inputs: [],
    mode: 'sdk-payloads',
    surface: 'file',
    direction: 'forward',
    windowSize: '0',
    bounds: 'live',
    mmapStrategy: 'buffer',
  });
  for (let i = 2; i < argv.length; i++) {
    const arg = argv[i];
    const next = () => {
      if (i + 1 >= argv.length) throw new Error(`missing value for ${arg}`);
      return argv[++i];
    };
    if (arg === '--input') args.inputs.push(next());
    else if (arg === '--mode') args.mode = next();
    else if (arg === '--surface') args.surface = next();
    else if (arg === '--direction') args.direction = next();
    else if (arg === '--window-size') args.windowSize = next();
    else if (arg === '--bounds') args.bounds = next();
    else if (arg === '--mmap-strategy') args.mmapStrategy = next();
    else throw new Error(`unknown argument: ${arg}`);
  }
  if (args.inputs.length === 0) throw new Error('missing --input');
  return args;
}

function stringifyResult(result) {
  const checksum = result.checksum;
  const json = JSON.stringify({ ...result, checksum: '__CHECKSUM__' });
  return json.replace('"__CHECKSUM__"', checksum.toString());
}

function run(args) {
  const statusBefore = processStatusKb();
  const started = performance.now();
  let counts;
  if (args.mode === 'sdk-entry' || args.mode === 'sdk-payloads') {
    counts = readSdk(args.inputs, args.surface, args.mode, args.direction);
  } else if (args.mode === 'facade-next' || args.mode === 'facade-data') {
    counts = readFacade(args.inputs, args.surface, args.mode, args.direction);
  } else {
    throw new Error(`invalid --mode for Node.js reader benchmark: ${args.mode}`);
  }
  const readSeconds = (performance.now() - started) / 1000;
  const statusAfter = processStatusKb();
  return { counts, readSeconds, statusBefore, statusAfter };
}

try {
  const args = parseArgs(process.argv);
  const { counts, readSeconds, statusBefore, statusAfter } = run(args);
  const result = {
    language: 'node',
    surface: args.surface,
    mode: args.mode,
    direction: args.direction,
    records: counts.records,
    fields: counts.fields,
    bytes: counts.bytes,
    checksum: counts.checksum,
    read_seconds: readSeconds,
    read_rows_per_second: readSeconds > 0 ? counts.records / readSeconds : 0.0,
    read_fields_per_second: readSeconds > 0 ? counts.fields / readSeconds : 0.0,
    read_bytes_per_second: readSeconds > 0 ? counts.bytes / readSeconds : 0.0,
    inputs: args.inputs,
    window_size: args.windowSize,
    bounds: args.bounds,
    mmap_strategy: args.mmapStrategy,
    timer_excludes: ['fixture generation', 'process startup', 'external verification'],
    process_status_before: statusBefore,
    process_status_after: statusAfter,
    errors: [],
  };
  process.stdout.write(stringifyResult(result) + '\n');
} catch (error) {
  process.stderr.write(`${error.stack || error.message}\n`);
  process.exitCode = 1;
}

File: experiments/python/cmd/reader_core_bench.py
#!/usr/bin/env python3
"""Python reader-core benchmark command."""

import argparse
import json
import sys
import time
from pathlib import Path


ROOT = Path(__file__).resolve().parents[2]
PYTHON_ROOT = ROOT / 'python'
sys.path.insert(0, str(PYTHON_ROOT))

from journal import (  # noqa: E402
    DirectoryReader,
    FileReader,
    ReaderOptions,
    READER_ACCESS_AUTO,
    READER_ACCESS_MMAP,
    SdJournalEnumerateAvailableData,
    SdJournalNext,
    SdJournalOpenDirectory,
    SdJournalOpenFiles,
    SdJournalRestartData,
)
from journal.reader_access import READER_ACCESS_READ_AT  # noqa: E402


class Counts:
    def __init__(self):
        self.records = 0
        self.fields = 0
        self.bytes = 0
        self.checksum = 0

    def add_payload(self, payload):
        self.fields += 1
        self.bytes += len(payload)
        self.checksum = checksum_payload(self.checksum, payload)

    def add_record_marker(self, value):
        self.records += 1
        self.checksum = rotate_left(self.checksum, 7) ^ int(value)


def rotate_left(value, shift):
    value &= 0xffffffffffffffff
    return ((value << shift) | (value >> (64 - shift))) & 0xffffffffffffffff


def checksum_payload(checksum, payload):
    checksum = rotate_left(checksum, 5) ^ len(payload)
    if payload:
        checksum ^= payload[0] << 8
        checksum ^= payload[-1]
    return checksum & 0xffffffffffffffff


def process_status_kb():
    try:
        status = Path('/proc/self/status').read_text(encoding='utf-8')
    except OSError:
        return {}
    wanted = {
        'VmSize', 'VmPeak', 'VmRSS', 'VmHWM', 'RssAnon', 'RssFile',
        'RssShmem', 'VmData', 'VmStk', 'VmExe', 'VmLib', 'VmPTE',
    }
    out = {}
    for line in status.splitlines():
        key, sep, value = line.partition(':')
        if not sep or key not in wanted:
            continue
        parts = value.split()
        if parts:
            try:
                out[f'{key}_kb'] = int(parts[0])
            except ValueError:
                continue
    return out


def advance(reader, direction):
    return reader.previous() if direction == 'backward' else reader.next()


def benchmark_reader_options(args):
    access_mode = benchmark_access_mode(args.mmap_strategy)
    kwargs = {
        'access_mode': access_mode,
        'bounds': args.bounds,
    }
    window_size = int(args.window_size)
    if window_size > 0:
        kwargs['window_size'] = window_size
    return ReaderOptions(**kwargs)


def benchmark_access_mode(strategy):
    normalized = str(strategy).lower().replace('_', '-')
    aliases = {
        'auto': READER_ACCESS_AUTO,
        'mmap': READER_ACCESS_MMAP,
        'windowed': READER_ACCESS_MMAP,
        'read-at': READER_ACCESS_READ_AT,
        'readat': READER_ACCESS_READ_AT,
        'pread': READER_ACCESS_READ_AT,
    }
    try:
        return aliases[normalized]
    except KeyError as err:
        raise ValueError(f'invalid Python mmap strategy: {strategy}') from err


def open_sdk_reader(inputs, surface, options):
    if surface == 'file':
        if len(inputs) != 1:
            raise ValueError('file surface requires exactly one --input')
        return FileReader.open(inputs[0], options=options)
    if surface == 'directory':
        if len(inputs) != 1:
            raise ValueError('directory surface requires exactly one --input')
        return DirectoryReader.open(inputs[0], options=options)
    if surface == 'open-files':
        return DirectoryReader.open_files(inputs, options=options)
    raise ValueError(f'invalid surface: {surface}')


def seek_reader(reader, direction):
    if direction == 'backward':
        reader.seek_tail()
    else:
        reader.seek_head()


def read_sdk(inputs, surface, mode, direction, options):
    reader = open_sdk_reader(inputs, surface, options)
    try:
        seek_reader(reader, direction)
        counts = Counts()
        while advance(reader, direction):
            if mode == 'sdk-entry':
                entry = reader.get_entry()
                counts.add_record_marker(entry['realtime'])
                for payload in entry['payloads']:
                    counts.add_payload(payload)
            elif mode == 'sdk-payloads':
                counts.add_record_marker(reader.get_realtime_usec())
                reader.visit_entry_payloads(counts.add_payload)
            else:
                raise ValueError(f'invalid SDK mode: {mode}')
        return counts, reader.access_stats()
    finally:
        reader.close()


def open_facade(inputs, surface, options):
    if surface == 'file' or surface == 'open-files':
        return SdJournalOpenFiles(inputs, 0, options=options)
    if surface == 'directory':
        if len(inputs) != 1:
            raise ValueError('directory surface requires exactly one --input')
        return SdJournalOpenDirectory(inputs[0], 0, options=options)
    raise ValueError(f'invalid facade surface: {surface}')


def read_facade(inputs, surface, mode, direction, options):
    journal = open_facade(inputs, surface, options)
    try:
        seek_reader(journal, direction)
        counts = Counts()
        while True:
            advanced = journal.previous() if direction == 'backward' else SdJournalNext(journal)
            if advanced == 0:
                break
            if mode == 'facade-next':
                counts.add_record_marker(journal.get_realtime_usec())
            elif mode == 'facade-data':
                counts.add_record_marker(journal.get_realtime_usec())
                SdJournalRestartData(journal)
                while True:
                    payload = SdJournalEnumerateAvailableData(journal)
                    if payload is None:
                        break
                    counts.add_payload(payload)
            else:
                raise ValueError(f'invalid facade mode: {mode}')
        return counts, journal.access_stats()
    finally:
        journal.close()


def run(args):
    options = benchmark_reader_options(args)
    status_before = process_status_kb()
    started = time.perf_counter()
    if args.mode in ('sdk-entry', 'sdk-payloads'):
        counts, access_stats = read_sdk(args.inputs, args.surface, args.mode, args.direction, options)
    elif args.mode in ('facade-next', 'facade-data'):
        counts, access_stats = read_facade(args.inputs, args.surface, args.mode, args.direction, options)
    else:
        raise ValueError(f'invalid --mode for Python reader benchmark: {args.mode}')
    elapsed = time.perf_counter() - started
    status_after = process_status_kb()
    return counts, access_stats, elapsed, status_before, status_after


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--input', dest='inputs', action='append', required=True)
    parser.add_argument('--mode', default='sdk-payloads')
    parser.add_argument('--surface', default='file')
    parser.add_argument('--direction', choices=('forward', 'backward'), default='forward')
    parser.add_argument('--window-size', default='0')
    parser.add_argument('--bounds', default='live')
    parser.add_argument('--mmap-strategy', default='mmap')
    args = parser.parse_args()

    counts, access_stats, read_seconds, status_before, status_after = run(args)
    print(json.dumps({
        'language': 'python',
        'surface': args.surface,
        'mode': args.mode,
        'direction': args.direction,
        'records': counts.records,
        'fields': counts.fields,
        'bytes': counts.bytes,
        'checksum': counts.checksum,
        'read_seconds': read_seconds,
        'read_rows_per_second': counts.records / read_seconds if read_seconds > 0 else 0.0,
        'read_fields_per_second': counts.fields / read_seconds if read_seconds > 0 else 0.0,
        'read_bytes_per_second': counts.bytes / read_seconds if read_seconds > 0 else 0.0,
        'inputs': args.inputs,
        'window_size': args.window_size,
        'bounds': args.bounds,
        'mmap_strategy': args.mmap_strategy,
        'access_stats': access_stats,
        'timer_excludes': ['fixture generation', 'process startup', 'external verification'],
        'process_status_before': status_before,
        'process_status_after': status_after,
        'errors': [],
    }))
    return 0


if __name__ == '__main__':
    raise SystemExit(main())
